File: app/utils/lang_utils.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayIncludes = ____lualib.__TS__ArrayIncludes
local __TS__ObjectKeys = ____lualib.__TS__ObjectKeys
local ____exports = {}
local curLang, langsData
local ____user_data = require("utils.user_data")
local get_string = ____user_data.get_string
local set = ____user_data.set
function ____exports.getSystemLang()
    local info = sys.get_sys_info()
    local code = info.language
    if __TS__ArrayIncludes({
        "ru",
        "be",
        "kk",
        "uk",
        "uz"
    }, code) then
        return "ru"
    else
        return code
    end
end
function ____exports.getAppLang()
    local lang = ____exports.getSystemLang()
    if not ____exports.hasLang(lang) then
        print("lang not found:", lang, "apply en")
        lang = "en"
    end
    return lang
end
function ____exports.hasLang(code)
    return langsData[code] ~= undefined
end
function ____exports.setLang(lang, saveLang)
    if saveLang == nil then
        saveLang = false
    end
    if saveLang then
        set("lang", lang)
    end
    curLang = lang
    ____exports.applyLang()
end
function ____exports.applyLang()
    local data = langsData[curLang]
    if not data then
        return
    end
    local keys = __TS__ObjectKeys(langsData[curLang])
    do
        local i = 0
        while i < #keys do
            local k = keys[i + 1]
            local val = data[k]
            local ok, node = pcall(gui.get_node, k)
            if ok then
                gui.set_text(node, val)
            end
            i = i + 1
        end
    end
end
curLang = "en"
langsData = {}
function ____exports.init()
    local saveLang = get_string("lang", "")
    if saveLang ~= "" then
        ____exports.setLang(saveLang)
    else
        ____exports.setLang(____exports.getAppLang())
    end
end
function ____exports.getCodeText(code)
    local data = langsData[curLang]
    if not data then
        print("not lang getCodeText:", curLang)
        return ""
    end
    if data[code] == undefined then
        print("код не найден", code)
        return ""
    end
    return data[code]
end
function ____exports.addLangData(lang, data)
    langsData[lang] = data
end
return ____exports

File: src/utils/lang_utils.ts
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable prefer-const */
/* eslint-disable @typescript-eslint/no-empty-function */

import { get_string, set } from "./user_data";

interface ILangData {
    [k: string]: string;
}

let curLang = 'en';
const langsData: { [k: string]: ILangData } = {};


export function init() {
    const saveLang = get_string('lang', '');
    if (saveLang != '')
        setLang(saveLang);
    else {
        setLang(getAppLang());
    }
}

export function getSystemLang() {
    const info = sys.get_sys_info();
    const code = info.language as string;
    if (['ru', 'be', 'kk', 'uk', 'uz'].includes(code))
        return 'ru';
    else
        return code;
}

export function getAppLang() {
    let lang = getSystemLang();
    if (!hasLang(lang)) {
        print('lang not found:', lang, 'apply en');
        lang = 'en';
    }
    return lang;
}

export function hasLang(code: string) {
    return langsData[code] !== undefined;
}

export function getCodeText(code: string) {
    const data = langsData[curLang];
    if (!data) {
        print('not lang getCodeText:', curLang);
        return '';
    }
    if (data[code] == undefined) {
        print('код не найден', code);
        return '';
    }
    return data[code];
}

export function addLangData(lang: string, data: ILangData) {
    langsData[lang] = data;
    //print('register lang:', lang);
}

export function setLang(lang: string, saveLang = false) {
    if (saveLang)
        set('lang', lang);
    curLang = lang;
    applyLang();
}


export function applyLang() {
    const data = langsData[curLang];
    if (!data)
        return;// print('not applyLang:', curLang);
    //print('ApplyLang:', curLang);
    const keys = Object.keys(langsData[curLang]);

    for (let i = 0; i < keys.length; i++) {
        const k = keys[i];
        const val = data[k];
        const [ok, node] = pcall(gui.get_node, k);
        if (ok)
            gui.set_text(node, val);
    }
}